<a href="https://colab.research.google.com/github/Epicencas/Lab1SL2/blob/main/Laboratorio1SL2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Laboratorio 1**

### Carga de Librerías

In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import roc_auc_score, roc_curve
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
from sklearn.model_selection import train_test_split
import keras
from keras.models import Sequential
from keras.layers import Dense

### Carga de los Datos ya tratados con Feature Engineering

In [2]:
data_path1 = '/content/drive/MyDrive/DataTratadaTrain.csv'
data_path2 = '/content/drive/MyDrive/DataTratadaTest.csv'
datatrain = pd.read_csv(data_path1)
datatest = pd.read_csv(data_path2)

### Definición de Valores den entrenamiento y de validación

In [3]:
X_train = datatrain.iloc[:, datatrain.columns != 'Survived']
y_train = datatrain.loc[:, 'Survived']
X_test = datatest.iloc[:, datatest.columns != 'Survived']


### Primera Estructura - Posición Kaggle: 14682

In [4]:
inputs = X_train.shape[1]

#base secuencial para definir la estructura de la Red (SIEMPRE INICIAR CON SQUENTIAL)
clasificador = Sequential()

#primera capa oclta
clasificador.add(Dense(input_dim = inputs, units=15, activation='relu', kernel_initializer='uniform'))

#segunda capa oculta
#clasificador.add(Dense(units=6, activation='relu', kernel_initializer='uniform'))

#capa de salida
clasificador.add(Dense(units=1, activation='sigmoid', kernel_initializer='uniform'))

In [5]:
#parametros de optimización
clasificador.compile(optimizer='SGD', loss='binary_crossentropy', metrics=['accuracy'])

#parametros de entrenamiento.
clasificador.fit(X_train, y_train, batch_size=25, epochs=550)

Epoch 1/550
36/36 [==============================] - 1s 3ms/step - loss: 1.0600 - accuracy: 0.5984
Epoch 2/550
36/36 [==============================] - 0s 4ms/step - loss: 0.6916 - accuracy: 0.6175
Epoch 3/550
36/36 [==============================] - 0s 3ms/step - loss: 0.6852 - accuracy: 0.6164
Epoch 4/550
36/36 [==============================] - 0s 5ms/step - loss: 0.6816 - accuracy: 0.6175
Epoch 5/550
36/36 [==============================] - 0s 4ms/step - loss: 0.6791 - accuracy: 0.6175
Epoch 6/550
36/36 [==============================] - 0s 3ms/step - loss: 0.6769 - accuracy: 0.6175
Epoch 7/550
36/36 [==============================] - 0s 3ms/step - loss: 0.6750 - accuracy: 0.6175
Epoch 8/550
36/36 [==============================] - 0s 3ms/step - loss: 0.6735 - accuracy: 0.6175
Epoch 9/550
36/36 [==============================] - 0s 4ms/step - loss: 0.6722 - accuracy: 0.6175
Epoch 10/550
36/36 [==============================] - 0s 5ms/step - loss: 0.6711 - accuracy: 0.6175
Epoch 11/

In [6]:
# Guardardamos la red.
path = '/content/drive/MyDrive/' 
clasificador.save(path + '6_6_sgd_sig_out.h5')

In [7]:
# Cargamos la red.
new_model = keras.models.load_model(path + '6_6_sgd_sig_out.h5')

In [8]:
#predicciones
y_preds = new_model.predict(X_test)
preds1 = pd.DataFrame(y_preds, columns = ['Survived'])
preds1.loc[preds1.Survived>50]='1'
preds1.loc[preds1.Survived<50]='0'
preds1.insert(0,'PassengerId',datatest['PassengerId'])
preds1.to_csv("Pred1.csv")

### Métricas Estructura 1

In [9]:
#Accuracy
y_preds = new_model.predict(X_train)
y_preds = (y_preds >= 0.5)
accuracy_score(y_train, y_preds)

0.6175478065241845

In [10]:
# ROC-AUC
roc_auc_score(y_train, y_preds, multi_class='ovr')

0.5

### Segunda Estructura - Posición Kaggle: 14682

In [11]:
inputs = X_train.shape[1]

#base secuencial para definir la estructura de la Red (SIEMPRE INICIAR CON SQUENTIAL)
clasificador = Sequential()

#primera capa oclta
clasificador.add(Dense(input_dim = inputs, units=6, activation='sigmoid', kernel_initializer='uniform'))

#segunda capa oculta
clasificador.add(Dense(units=6, activation='sigmoid', kernel_initializer='uniform'))

#capa de salida
clasificador.add(Dense(units=1, activation='sigmoid', kernel_initializer='uniform'))

In [12]:
#parametros de optimización
clasificador.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

#parametros de entrenamiento.
clasificador.fit(X_train, y_train, batch_size=30, epochs=1000)

Epoch 1/1000
30/30 [==============================] - 1s 2ms/step - loss: 0.6853 - accuracy: 0.6175
Epoch 2/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6778 - accuracy: 0.6175
Epoch 3/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6724 - accuracy: 0.6175
Epoch 4/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6686 - accuracy: 0.6175
Epoch 5/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6653 - accuracy: 0.6175
Epoch 6/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6624 - accuracy: 0.6175
Epoch 7/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6593 - accuracy: 0.6175
Epoch 8/1000
30/30 [==============================] - 0s 3ms/step - loss: 0.6565 - accuracy: 0.6175
Epoch 9/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6544 - accuracy: 0.6175
Epoch 10/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6515 - accuracy: 0.6175

In [13]:
# Guardardamos la red.
path = '/content/drive/MyDrive/' 
clasificador.save(path + '6_6_adam_sig_out.h5')

In [14]:
# Cargamos la red.
new_model = keras.models.load_model(path + '6_6_adam_sig_out.h5')

In [15]:
#predicciones
#predicciones
y_preds = new_model.predict(X_test)
preds2 = pd.DataFrame(y_preds, columns = ['Survived'])
preds2.loc[preds2.Survived>50]='1'
preds2.loc[preds2.Survived<50]='0'
preds2.insert(0,'PassengerId',datatest['PassengerId'])
preds2.to_csv("Pred2.csv")

### Métricas Estructura 2

In [16]:
#Accuracy
y_preds = new_model.predict(X_train)
y_preds = (y_preds >= 0.5)
accuracy_score(y_train, y_preds)

0.7761529808773904

In [17]:
# ROC-AUC
roc_auc_score(y_train, y_preds, multi_class='ovr')

0.7219088181720775

### Tercera Estructura - Posición Kaggle: 14682

In [18]:
inputs = X_train.shape[1]

#base secuencial para definir la estructura de la Red (SIEMPRE INICIAR CON SQUENTIAL)
clasificador = Sequential()

#primera capa oclta
clasificador.add(Dense(input_dim = inputs, units=10, activation='sigmoid', kernel_initializer='uniform'))

#segunda capa oculta
#clasificador.add(Dense(units=10, activation='sigmoid', kernel_initializer='uniform'))

#capa de salida
clasificador.add(Dense(units=1, activation='sigmoid', kernel_initializer='uniform'))

In [19]:
#parametros de optimización
clasificador.compile(optimizer='RMSprop', loss='binary_crossentropy', metrics=['accuracy'])

#parametros de entrenamiento.
clasificador.fit(X_train, y_train, batch_size=30, epochs=500)

Epoch 1/500
30/30 [==============================] - 1s 2ms/step - loss: 0.6927 - accuracy: 0.5399
Epoch 2/500
30/30 [==============================] - 0s 2ms/step - loss: 0.6786 - accuracy: 0.7120
Epoch 3/500
30/30 [==============================] - 0s 2ms/step - loss: 0.6644 - accuracy: 0.7267
Epoch 4/500
30/30 [==============================] - 0s 2ms/step - loss: 0.6524 - accuracy: 0.7188
Epoch 5/500
30/30 [==============================] - 0s 2ms/step - loss: 0.6382 - accuracy: 0.7357
Epoch 6/500
30/30 [==============================] - 0s 2ms/step - loss: 0.6274 - accuracy: 0.7289
Epoch 7/500
30/30 [==============================] - 0s 2ms/step - loss: 0.6148 - accuracy: 0.7402
Epoch 8/500
30/30 [==============================] - 0s 2ms/step - loss: 0.6045 - accuracy: 0.7435
Epoch 9/500
30/30 [==============================] - 0s 2ms/step - loss: 0.5954 - accuracy: 0.7323
Epoch 10/500
30/30 [==============================] - 0s 2ms/step - loss: 0.5869 - accuracy: 0.7447
Epoch 11/

In [20]:
# Guardardamos la red.
path = '/content/drive/MyDrive/' 
clasificador.save(path + '10_10_RMSprop_sig_out.h5')

In [21]:
# Cargamos la red.
new_model = keras.models.load_model(path + '10_10_RMSprop_sig_out.h5')

In [22]:
#predicciones
y_preds = new_model.predict(X_test)
preds3 = pd.DataFrame(y_preds, columns = ['Survived'])
preds3.loc[preds3.Survived>50]='1'
preds3.loc[preds3.Survived<50]='0'
preds3.insert(0,'PassengerId',datatest['PassengerId'])
preds3.to_csv("Pred3.csv")

### Métricas Estructura 3

In [23]:
#Accuracy
y_preds = new_model.predict(X_train)
y_preds = (y_preds >= 0.5)
accuracy_score(y_train, y_preds)

0.7952755905511811

In [24]:
# ROC-AUC
roc_auc_score(y_train, y_preds, multi_class='ovr')

0.7793796207007394

### Estructura Cuatro - Posición Kaggle: 14682

In [25]:
inputs = X_train.shape[1]

#base secuencial para definir la estructura de la Red (SIEMPRE INICIAR CON SQUENTIAL)
clasificador = Sequential()

#primera capa oclta
clasificador.add(Dense(input_dim = inputs, units=3, activation='sigmoid', kernel_initializer='uniform'))

#segunda capa oculta
clasificador.add(Dense(units=6, activation='sigmoid', kernel_initializer='uniform'))

#tercera capa oculta
clasificador.add(Dense(units=6, activation='sigmoid', kernel_initializer='uniform'))

#capa de salida
clasificador.add(Dense(units=1, activation='sigmoid', kernel_initializer='uniform'))

In [26]:
#parametros de optimización
clasificador.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

#parametros de entrenamiento.
clasificador.fit(X_train, y_train, batch_size=20, epochs=500)

Epoch 1/500
45/45 [==============================] - 1s 2ms/step - loss: 0.6833 - accuracy: 0.6175
Epoch 2/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6742 - accuracy: 0.6175
Epoch 3/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6693 - accuracy: 0.6175
Epoch 4/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6665 - accuracy: 0.6175
Epoch 5/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6651 - accuracy: 0.6175
Epoch 6/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6644 - accuracy: 0.6175
Epoch 7/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6642 - accuracy: 0.6175
Epoch 8/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6636 - accuracy: 0.6175
Epoch 9/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6633 - accuracy: 0.6175
Epoch 10/500
45/45 [==============================] - 0s 3ms/step - loss: 0.6631 - accuracy: 0.6175
Epoch 11/

In [27]:
# Guardardamos la red.
path = '/content/drive/MyDrive/' 
clasificador.save(path + '3_6_adam_sig_out.h5')

In [28]:
# Cargamos la red.
new_model = keras.models.load_model(path + '3_6_adam_sig_out.h5')

In [29]:
#predicciones
y_preds = new_model.predict(X_test)
preds4 = pd.DataFrame(y_preds, columns = ['Survived'])
preds4.loc[preds4.Survived>50]='1'
preds4.loc[preds4.Survived<50]='0'
preds4.insert(0,'PassengerId',datatest['PassengerId'])
preds4.to_csv("Pred4.csv")

### Métricas Estructura 4

In [30]:
#Accuracy
y_preds = new_model.predict(X_train)
y_preds = (y_preds >= 0.5)
accuracy_score(y_train, y_preds)

0.749156355455568

In [31]:
# ROC-AUC
roc_auc_score(y_train, y_preds, multi_class='ovr')

0.7319618557805635

### Estructura Cinco - Posición Kaggle: 14682

In [32]:
inputs = X_train.shape[1]

#base secuencial para definir la estructura de la Red (SIEMPRE INICIAR CON SQUENTIAL)
clasificador = Sequential()

#primera capa oclta
clasificador.add(Dense(input_dim = inputs, units=3, activation='sigmoid', kernel_initializer='uniform'))

#segunda capa oculta
clasificador.add(Dense(units=3, activation='sigmoid', kernel_initializer='uniform'))

#capa de salida
clasificador.add(Dense(units=1, activation='sigmoid', kernel_initializer='uniform'))

In [33]:
#parametros de optimización
clasificador.compile(optimizer='Adamax', loss='binary_crossentropy', metrics=['accuracy'])

#parametros de entrenamiento.
clasificador.fit(X_train, y_train, batch_size=20, epochs=500)

Epoch 1/500
45/45 [==============================] - 1s 2ms/step - loss: 0.6896 - accuracy: 0.6175
Epoch 2/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6857 - accuracy: 0.6175
Epoch 3/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6826 - accuracy: 0.6175
Epoch 4/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6798 - accuracy: 0.6175
Epoch 5/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6775 - accuracy: 0.6175
Epoch 6/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6753 - accuracy: 0.6175
Epoch 7/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6737 - accuracy: 0.6175
Epoch 8/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6722 - accuracy: 0.6175
Epoch 9/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6709 - accuracy: 0.6175
Epoch 10/500
45/45 [==============================] - 0s 2ms/step - loss: 0.6696 - accuracy: 0.6175
Epoch 11/

In [34]:
# Guardardamos la red.
path = '/content/drive/MyDrive/' 
clasificador.save(path + '3_3_adamax_sig_out.h5')

In [35]:
# Cargamos la red.
new_model = keras.models.load_model(path + '3_3_adamax_sig_out.h5')

In [36]:
#predicciones
y_preds = new_model.predict(X_test)
preds5 = pd.DataFrame(y_preds, columns = ['Survived'])
preds5.loc[preds5.Survived>50]='1'
preds5.loc[preds5.Survived<50]='0'
preds5.insert(0,'PassengerId',datatest['PassengerId'])
preds5.to_csv("Pred5.csv")

### Métricas Estructura 5

In [37]:
#Accuracy
y_preds = new_model.predict(X_train)
y_preds = (y_preds >= 0.5)
accuracy_score(y_train, y_preds)

0.7930258717660292

In [38]:
# ROC-AUC
roc_auc_score(y_train, y_preds, multi_class='ovr')

0.76300225008036

### Estructura Seis 14682

In [39]:
inputs = X_train.shape[1]

#base secuencial para definir la estructura de la Red (SIEMPRE INICIAR CON SQUENTIAL)
clasificador = Sequential()

#primera capa oclta
clasificador.add(Dense(input_dim = inputs, units=3, activation='sigmoid', kernel_initializer='uniform'))

#segunda capa oculta
clasificador.add(Dense(units=3, activation='sigmoid', kernel_initializer='uniform'))

#capa de salida
clasificador.add(Dense(units=1, activation='sigmoid', kernel_initializer='uniform'))

In [40]:
#parametros de optimización
clasificador.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

#parametros de entrenamiento.
clasificador.fit(X_train, y_train, batch_size=40, epochs=2000)

Epoch 1/2000
23/23 [==============================] - 1s 2ms/step - loss: 0.6887 - accuracy: 0.6175
Epoch 2/2000
23/23 [==============================] - 0s 2ms/step - loss: 0.6839 - accuracy: 0.6175
Epoch 3/2000
23/23 [==============================] - 0s 2ms/step - loss: 0.6796 - accuracy: 0.6175
Epoch 4/2000
23/23 [==============================] - 0s 2ms/step - loss: 0.6761 - accuracy: 0.6175
Epoch 5/2000
23/23 [==============================] - 0s 2ms/step - loss: 0.6727 - accuracy: 0.6175
Epoch 6/2000
23/23 [==============================] - 0s 2ms/step - loss: 0.6699 - accuracy: 0.6175
Epoch 7/2000
23/23 [==============================] - 0s 2ms/step - loss: 0.6676 - accuracy: 0.6175
Epoch 8/2000
23/23 [==============================] - 0s 2ms/step - loss: 0.6655 - accuracy: 0.6175
Epoch 9/2000
23/23 [==============================] - 0s 2ms/step - loss: 0.6637 - accuracy: 0.6175
Epoch 10/2000
23/23 [==============================] - 0s 2ms/step - loss: 0.6618 - accuracy: 0.6175

In [41]:
# Guardardamos la red.
path = '/content/drive/MyDrive/' 
clasificador.save(path + '3_3_adam_sig_out.h5')

In [42]:
# Cargamos la red.
new_model = keras.models.load_model(path + '3_3_adamax_sig_out.h5')

In [43]:
#predicciones
y_preds = new_model.predict(X_test)
preds6 = pd.DataFrame(y_preds, columns = ['Survived'])
preds6.loc[preds6.Survived>50]='1'
preds6.loc[preds6.Survived<50]='0'
preds6.insert(0,'PassengerId',datatest['PassengerId'])
preds6.to_csv("Pred6.csv")

### Metricas Estructura 6

In [44]:
#Accuracy
y_preds = new_model.predict(X_train)
y_preds = (y_preds >= 0.5)
accuracy_score(y_train, y_preds)

0.7930258717660292

In [45]:
# ROC-AUC
roc_auc_score(y_train, y_preds, multi_class='ovr')

0.76300225008036

### Estructura Siete 14682

In [46]:
inputs = X_train.shape[1]

#base secuencial para definir la estructura de la Red (SIEMPRE INICIAR CON SQUENTIAL)
clasificador = Sequential()

#primera capa oclta
clasificador.add(Dense(input_dim = inputs, units=4, activation='sigmoid', kernel_initializer='uniform'))

#segunda capa oculta
#clasificador.add(Dense(units=6, activation='sigmoid', kernel_initializer='uniform'))

#capa de salida
clasificador.add(Dense(units=1, activation='sigmoid', kernel_initializer='uniform'))

In [47]:
#parametros de optimización
clasificador.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

#parametros de entrenamiento.
clasificador.fit(X_train, y_train, batch_size=30, epochs=1000)

Epoch 1/1000
30/30 [==============================] - 1s 3ms/step - loss: 0.6868 - accuracy: 0.6085
Epoch 2/1000
30/30 [==============================] - 0s 3ms/step - loss: 0.6808 - accuracy: 0.6175
Epoch 3/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6757 - accuracy: 0.6175
Epoch 4/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6711 - accuracy: 0.6175
Epoch 5/1000
30/30 [==============================] - 0s 3ms/step - loss: 0.6662 - accuracy: 0.6175
Epoch 6/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6614 - accuracy: 0.6175
Epoch 7/1000
30/30 [==============================] - 0s 3ms/step - loss: 0.6572 - accuracy: 0.6175
Epoch 8/1000
30/30 [==============================] - 0s 3ms/step - loss: 0.6531 - accuracy: 0.6175
Epoch 9/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6497 - accuracy: 0.6175
Epoch 10/1000
30/30 [==============================] - 0s 2ms/step - loss: 0.6472 - accuracy: 0.6175

In [48]:
# Guardardamos la red.
path = '/content/drive/MyDrive/' 
clasificador.save(path + '3_3_adam_sig_out.h5')

In [49]:
# Cargamos la red.
new_model = keras.models.load_model(path + '3_6_adam_sig_out.h5')

In [50]:
#predicciones
#predicciones
y_preds = new_model.predict(X_test)
preds7 = pd.DataFrame(y_preds, columns = ['Survived'])
preds7.loc[preds7.Survived>50]='1'
preds7.loc[preds7.Survived<50]='0'
preds7.insert(0,'PassengerId',datatest['PassengerId'])
preds7.to_csv("Pred7.csv")

### Metricas Estructura 7

In [51]:
#Accuracy
y_preds = new_model.predict(X_train)
y_preds = (y_preds >= 0.5)
accuracy_score(y_train, y_preds)

0.749156355455568

In [52]:
# ROC-AUC
roc_auc_score(y_train, y_preds, multi_class='ovr')

0.7319618557805635

### Estructura Ocho - Posición Kaggle: 14682

In [53]:
#KFolds
def red_v5():
  #inputs = X_train.shape[1]

  #base secuencial para definir la estructura de la Red (SIEMPRE INICIAR CON SQUENTIAL)
  clasificador = Sequential()

  #primera capa oclta
  clasificador.add(Dense(input_dim = 8, units=10, activation='sigmoid', kernel_initializer='uniform'))

  #segunda capa oculta
  clasificador.add(Dense(units=10, activation='sigmoid', kernel_initializer='uniform'))

  #capa de salida
  clasificador.add(Dense(units=1, activation='sigmoid', kernel_initializer='uniform'))

  #parametros de optimización
  clasificador.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
  
  return clasificador

In [54]:
clasificador = KerasClassifier(build_fn=red_v5, batch_size=30, nb_epoch=1200)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  """Entry point for launching an IPython kernel.


In [55]:
kfolds = KFold(n_splits=10, shuffle=True)
accs = cross_val_score(clasificador, X=X_train, y=y_train, cv=kfolds, verbose=1, n_jobs=-1)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:   11.4s finished


In [56]:
accs

array([0.67415732, 0.62921351, 0.74157304, 0.53932583, 0.57303369,
       0.52808988, 0.58426964, 0.56179774, 0.73033708, 0.61363637])

### Conclusión

La mejor estructura es la 2 con 1 sola capa y 10 neuronas, optimizador RMSprop y activador sigmoide. Es la mejor ya que es la que logró el accuracy y el ROCAUC más altos, 0.79 y 0.76 respectivamente.